File: src/jvmTest/kotlin/community/flock/kotlinx/rgxgen/config/WhitespaceConfigurationTests.java
package community.flock.kotlinx.rgxgen.config;

import community.flock.kotlinx.rgxgen.RgxGen;
import community.flock.kotlinx.rgxgen.model.WhitespaceChar;
import community.flock.kotlinx.rgxgen.util.Util;
import kotlin.random.Random;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.regex.Pattern;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class WhitespaceConfigurationTests {
    public static final int COUNT_OF_ITERATIONS       = 1000;
    public static final int LARGE_COUNT_OF_ITERATIONS = 1000;

    @Test
    void byDefaultOnlySpaceAndTabIsGeneratedTest() {
        Random random = Util.newRandom(100500);
        Pattern pattern = Pattern.compile("[ \t]");
        for (int i = 0; i < COUNT_OF_ITERATIONS; i++) {
            String generated = RgxGen.parse("\\s").generate(random);
            assertTrue(pattern.matcher(generated).matches(), "Failed for value '" + generated + '\'');
        }
    }

    @Test
    void configuredWhitespaceGenerationTest() {
        RgxGenProperties properties = new RgxGenProperties();
        RgxGenOption.WHITESPACE_DEFINITION.setInProperties(properties, Arrays.asList(WhitespaceChar.FORM_FEED, WhitespaceChar.VERTICAL_TAB));
        Pattern pattern = Pattern.compile("[\u000B\f]");
        Random random = Util.newRandom(100500);
        for (int i = 0; i < COUNT_OF_ITERATIONS; i++) {
            String generated = RgxGen.parse(properties, "\\s").generate(random);
            assertTrue(pattern.matcher(generated).matches(), "Failed for value '" + generated + '\'');
        }
    }

    @Test
    void notMatchingGenerationAlwaysExcludesAllWhitespacesTest() {
        Random random = Util.newRandom(100500);
        Pattern pattern = Pattern.compile("\\s");
        for (int i = 0; i < LARGE_COUNT_OF_ITERATIONS; i++) {
            String generated = RgxGen.parse("\\s").generateNotMatching(random);
            assertFalse(pattern.matcher(generated).matches(), "Failed for value '" + generated + '\'');
        }
    }

    @Test
    void notWhitespaceCharacterAlwaysExcludesAllWhitespacesTest() {
        Random random = Util.newRandom(100500);
        Pattern pattern = Pattern.compile("\\s");
        for (int i = 0; i < LARGE_COUNT_OF_ITERATIONS; i++) {
            String generated = RgxGen.parse("\\S").generate(random);
            assertFalse(pattern.matcher(generated).matches(), "Failed for value '" + generated + '\'');
        }
    }

    @Test
    void matchingGenerationWithinSquareBracketsTest() {
        Random random = Util.newRandom(100500);
        Pattern pattern = Pattern.compile("[a \t]");
        for (int i = 0; i < COUNT_OF_ITERATIONS; i++) {
            String generated = RgxGen.parse("[a\\s]").generate(random);
            assertTrue(pattern.matcher(generated).matches(), "Failed for value '" + generated + '\'');
        }
    }

    @Test
    void notMatchingGenerationWithinSquareBracketsTest() {
        Random random = Util.newRandom(100500);
        Pattern pattern = Pattern.compile("[^a\\s]");
        for (int i = 0; i < COUNT_OF_ITERATIONS; i++) {
            String generated = RgxGen.parse("[a\\s]").generateNotMatching(random);
            assertTrue(pattern.matcher(generated).matches(), "Failed for value '" + generated + '\'');
        }
    }

    @Test
    void matchingGenerationNegativeWithinSquareBracketsTest() {
        Random random = Util.newRandom(100500);
        Pattern pattern = Pattern.compile("[^a\\s]");
        for (int i = 0; i < COUNT_OF_ITERATIONS; i++) {
            String generated = RgxGen.parse("[^a\\s]").generate(random);
            assertTrue(pattern.matcher(generated).matches(), "Failed for value '" + generated + '\'');
        }
    }
}

File: src/jvmTest/kotlin/community/flock/kotlinx/rgxgen/config/WhitespaceConfigurationTests.kt
package community.flock.kotlinx.rgxgen.config

import community.flock.kotlinx.rgxgen.RgxGen.Companion.parse
import community.flock.kotlinx.rgxgen.model.WhitespaceChar
import community.flock.kotlinx.rgxgen.util.Util.newRandom
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.util.Arrays
import java.util.regex.Pattern

class WhitespaceConfigurationTests {
    @Test
    fun byDefaultOnlySpaceAndTabIsGeneratedTest() {
        val random = newRandom(100500)
        val pattern = Pattern.compile("[ \t]")
        for (i in 0 until COUNT_OF_ITERATIONS) {
            val generated = parse("\\s").generate(random)
            Assertions.assertTrue(pattern.matcher(generated).matches(), "Failed for value '$generated'")
        }
    }

    @Test
    fun configuredWhitespaceGenerationTest() {
        val properties: RgxGenProperties<Any> = RgxGenProperties()
        RgxGenOption.WHITESPACE_DEFINITION.setInProperties(
            properties,
            Arrays.asList(WhitespaceChar.FORM_FEED, WhitespaceChar.VERTICAL_TAB)
        )
        val pattern = Pattern.compile("[\u000B\u000c]")
        val random = newRandom(100500)
        for (i in 0 until COUNT_OF_ITERATIONS) {
            val generated = parse(properties, "\\s").generate(random)
            Assertions.assertTrue(pattern.matcher(generated).matches(), "Failed for value '$generated'")
        }
    }

    @Test
    fun notMatchingGenerationAlwaysExcludesAllWhitespacesTest() {
        val random = newRandom(100500)
        val pattern = Pattern.compile("\\s")
        for (i in 0 until LARGE_COUNT_OF_ITERATIONS) {
            val generated = parse("\\s").generateNotMatching(random)
            Assertions.assertFalse(pattern.matcher(generated).matches(), "Failed for value '$generated'")
        }
    }

    @Test
    fun notWhitespaceCharacterAlwaysExcludesAllWhitespacesTest() {
        val random = newRandom(100500)
        val pattern = Pattern.compile("\\s")
        for (i in 0 until LARGE_COUNT_OF_ITERATIONS) {
            val generated = parse("\\S").generate(random)
            Assertions.assertFalse(pattern.matcher(generated).matches(), "Failed for $i value '$generated'")
        }
    }

    @Test
    fun matchingGenerationWithinSquareBracketsTest() {
        val random = newRandom(100500)
        val pattern = Pattern.compile("[a \t]")
        for (i in 0 until COUNT_OF_ITERATIONS) {
            val generated = parse("[a\\s]").generate(random)
            Assertions.assertTrue(pattern.matcher(generated).matches(), "Failed for $i value '$generated'")
        }
    }

    @Test
    fun notMatchingGenerationWithinSquareBracketsTest() {
        val random = newRandom(100500)
        val pattern = Pattern.compile("[^a\\s]")
        for (i in 0 until COUNT_OF_ITERATIONS) {
            val generated = parse("[a\\s]").generateNotMatching(random)
            Assertions.assertTrue(pattern.matcher(generated).matches(), "Failed for $i value '$generated'")
        }
    }

    @Test
    fun matchingGenerationNegativeWithinSquareBracketsTest() {
        val random = newRandom(100500)
        val pattern = Pattern.compile("[^a\\s]")
        for (i in 0 until COUNT_OF_ITERATIONS) {
            val generated = parse("[^a\\s]").generate(random)
            Assertions.assertTrue(pattern.matcher(generated).matches(), "Failed for $i value '$generated'")
        }
    }

    companion object {
        const val COUNT_OF_ITERATIONS: Int = 1000
        const val LARGE_COUNT_OF_ITERATIONS: Int = 1000
    }
}
